fix: Popover closes on change of dropdowns values

geido opened the pull request:
### SUMMARY
The option `menuPortalTarget` of the react-select component was introduced in order for the dropdowns to show fully when overlaying certain container elements. However, setting the `menuPortalTarget` to `document.body` (as originally the multi-choice dropdown did and later the single-choice) comes with several issues, such as the click events being targeted outside of the container elements, that ultimately caused this issue #12388.

This PR removes the `menuPortalTarget` for both single-choice and multi-choice dropdowns. In addition to that, this introduces the optional ability to force the overflow visibility of the dropdowns for cases where setting the `overflow` `visible` isn't viable on the container element (see for instance issue #12004 )

It also adjusts some excessive z-index by setting it high enough to give it precedence over the Antd Popover and other components.

Closes #12388 

### BEFORE

https://user-images.githubusercontent.com/60598000/104368923-06b6fc00-551d-11eb-935b-b6b7ba657dc3.mp4

### AFTER

https://user-images.githubusercontent.com/60598000/104368799-dc653e80-551c-11eb-9778-3fed7cede574.mp4



### TEST PLAN
1. Open the single and multi-choice dropdowns throughout the project
2. Make sure the dropdowns are displaying correctly

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: #12388 
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

geido committed:
Set menuPortalTarget to nearest parent - tentative

geido committed:
Merge remote-tracking branch 'upstream/master' into fix/issue_12388_popover_closes_onchange

geido commented:
@junlincc @rusackas

geido committed:
Merge remote-tracking branch 'upstream/master' into fix/issue_12388_popover_closes_onchange

geido committed:
Fix lint issue

junlincc commented:
@villebro @rusackas @ktmud need to make sure merging this PR will not create any conflict to changes in #12406

adam-stasiak commented:
I found issue related to scrolling with dropdown active:
Left side - PR, before change
https://user-images.githubusercontent.com/25153919/104233738-ab203c00-5452-11eb-9f21-908329b3de88.mov

geido commented:
For reference, this is the react-select issue related to scroll https://github.com/JedWatson/react-select/issues/4088

ktmud commented:
Can we add an option to `SupersetSelect` to disable the portal wrap by default and only enable it for places where it's likely to overflow the parent container?

kkucharc commented on superset-frontend/src/components/Select/SupersetStyledSelect.tsx:
```diff
@@ -259,36 +257,43 @@ function styled<
-        components={components}
-        {...restProps}
-      />
+      <div ref={menuPortalTargetRef}>
```
I've tried to test it and I can't find places where exact benefit of referring `menuPortalTarget` to the div. Maybe you could provide some gif or screens where it helps? Otherwise maybe `menuPortalTarget={document.body}` would be enough?

geido replied:
Hi. @kkucharc thanks for your feedback. The `menuPortalTarget` to `document.body` will create the issue #12388 as the click event will be targeted outside of the Popover container. This is actually one of the main reasons why I am implementing this change. However, I think we can make this optional and just use it as a fallback, while managing all the other cases individually with ad hoc solutions where possible. I am looking into that.

kkucharc reviewed (commented):
Thank you for the contribution! Seems that moving `menuPortalTarget` higher structure may help.

I also saw scroll problem that @adam-stasiak mentioned. Unfortunately it happens in many places such as explore:
![Large GIF (798x774)](https://user-images.githubusercontent.com/2536609/104335058-5385dd00-54f3-11eb-9204-ddd251590ac0.gif)

geido committed:
Merge remote-tracking branch 'upstream/master' into fix/issue_12388_popover_closes_onchange

geido committed:
Add optional forceOverflow option

geido committed:
Merge remote-tracking branch 'upstream/master' into fix/issue_12388_popover_closes_onchange

geido committed:
Remove unnecessary menuPortalTarget

geido committed:
Delete yarn.lock

geido commented:
I have cleaned up the things a bit. The `menuPortalTarget` wasn't really needed and I removed it. What is needed is only the `menuPosition` to be `fixed` for instances where setting the overflow to visible on the container element isn't possible. I made that optional as well by introducing a `forceOverflow` option so that individual cases can be handle appropriately. Finally, when the option`forceOverflow` is used, I am closing the dropdown on scroll when the target of the scroll isn't the dropdown itself so that issue mentioned by @adam-stasiak and @kkucharc is mitigated.

geido committed:
Force overflow dropdown refresh-interval modal

geido committed:
Merge branch 'fix/issue_12388_popover_closes_onchange' of github.com:geido/incubator-superset into fix/issue_12388_popover_closes_onchange

geido committed:
Merge remote-tracking branch 'upstream/master' into fix/issue_12388_popover_closes_onchange

adam-stasiak commented:
Removing data from dropdown causes modal dismiss.

Uploading Nagranie z ekranu 2021-01-12 o 22.51.08.mov…

geido commented:
@adam-stasiak seems related to a different issue to me

kkucharc commented on superset-frontend/src/components/Select/SupersetStyledSelect.tsx:
```diff
@@ -243,6 +244,18 @@ function styled<
 
+    // handle forcing dropdown overflow
+    // use only when setting overflow:visible isn't possible on the container element
+    if (forceOverflow) {
```
Are there any places that have this option already as `true`?

geido replied:
Yes you can see two instances in this PR already. One is on the dashboard filters and the other one is used for the refresh-interval modal. The one for the dashboard filters is required. The one for refresh-interval is more to show how it looks like on a modal but it is not necessary and can be removed.

kkucharc replied:
I see, thanks for explanation 👍

kkucharc reviewed (commented):
Thanks a lot for improvement!
According to the video in your comment: I understand that keeping dropdown open and sticked to the input until we click somewhere else is just impossible to achieve?

geido commented:
> Thanks a lot for improvement!
> According to the video in your comment: I understand that keeping dropdown open and sticked to the input until we click somewhere else is just impossible to achieve?

@kkucharc impossible is not I would define it. The problem is this is a known issue of the react-select library and it does not have a solution right now. Fixing this problem on our side might mean coming up with hacks that are not desirable. Closing the dropdown on scroll is already some sort of hack, but I would not go beyond that really.

geido commented:
@kkucharc that's an issue affecting all the modals. I think I have a solution for that and I am going to submit a separate PR. In fact, we don't need to necessarily use the `forceOverflow` option to fix that problem. In general, I'd use the `forceOverflow` option introduced by this PR only for those cases where the overflow issue can't be solved otherwise.

adam-stasiak commented:
Tested manually across other places in app - looks fine.

ktmud reviewed (approved):
LGTM, tested and works great after rebase.

ktmud commented:
@junlincc @villebro can we make sure this PR goes into `v1.0`, too? Currently in `1.0.0 RC2`, the dropdown menu of react-select is sticky on page scroll, which is not very nice:

<img src="https://user-images.githubusercontent.com/335541/104633922-bd69c800-5654-11eb-9c69-906148863e3d.gif" width="600" alt="sticky menu">

junlincc commented:
we are canceling rc2 kicking of rc3 to include this and a couple of others @ktmud

villebro commented:
@ktmud I already cancelled rc2, so I can pick this into rc3. @junlincc remember to flag all PRs you want included in 1.0 with the appropriate label so I have a checklist.
